libcst/codemod/commands/rename.py

```diff
@@ -0,0 +1,192 @@
+
+        parser.add_argument(
+            "--new_name",
+            dest="new_name",
+            required=True,
+            help="Full dotted name of replacement object. Eg: `foo.bar.baz`",
+        )
+
+    def __init__(self, context: CodemodContext, old_name: str, new_name: str) -> None:
+        super().__init__(context)
+        self.old_name = old_name
+        self.new_name = new_name
+
+        old_names = self.old_name.rpartition(".")
+        new_names = self.new_name.rpartition(".")
+
+        # Mapping for easy lookup later.
+        self.equivalents: Dict[str, str] = {
+            old_names[2]: new_names[2],
+            "".join(old_names[:2]): new_names[0],
+        }
+
+        self.scheduled_imports = False
+
+    def visit_Import(self, node: cst.Import) -> None:
+        for import_alias in node.names:
+            alias_name = get_full_name_for_node(import_alias.name)
+            if alias_name is not None:
+                self.record_asname(import_alias, alias_name)
+
+                # If the import statement is exactly equivalent to the old name, replace it.
+                if alias_name == self.old_name:
+                    self.cleanup_imports(node=node, new_module=self.new_name)
+
+    def leave_ImportFrom(
+        self, original_node: cst.ImportFrom, updated_node: cst.ImportFrom
+    ) -> cst.ImportFrom:
+        module = updated_node.module
+        if module is not None:
+            imported_module_name = get_full_name_for_node(module)
+            names = original_node.names
+
+            if imported_module_name is None or not isinstance(names, Sequence):
+                return updated_node
+            elif imported_module_name == self.old_name:
+                # Simply rename the imported module on the spot.
+                return updated_node.with_changes(
+                    module=cst.parse_expression(self.new_name)
+                )
+            else:
+                new_names = []
+                for import_alias in names:
+                    alias_name = get_full_name_for_node(import_alias.name)
+                    if alias_name is not None:
+                        qual_name = f"{imported_module_name}.{alias_name}"
+                        if self.old_name == qual_name:
+                            self.record_asname(import_alias, alias_name)
+
+                            replacement_module = self.gen_replacement(
+                                imported_module_name + "."
+                            )
+                            replacement_obj = self.gen_replacement(alias_name)
+                            new_import_alias_name: cst.BaseExpression = cst.parse_expression(
+                                replacement_obj
+                            )
+                            if not isinstance(
+                                new_import_alias_name, (cst.Attribute, cst.Name)
+                            ):
+                                raise Exception("Something went wrong!")
```

Comment:
This is a pretty classic not-useful error message :) I understand it should never happen and you're just keeping the type-checker happy, but it still seems like nothing would be lost by changing it to `"parse_expression() on dotted path returned non-Attribute-or-Name"`.

Another alternative to consider might be a simple `assert`: `assert isinstance(new_import_alias_name, (cst.Attribute, cst.Name))`